### PR TITLE
fix: leave balance reports

### PR DESCRIPTION
**Pull request by Mangesh-Khairnar**
Fixes
- opening leave balance was fetching the leave allocation. This change fetches the leave balance based on the start date from the filter.
- Checks for leave approvers in employee leave balance summary and display only those records for which user is a leave approver(department approver).

### CODE REVIEW

File: erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py [hunk 1]
```diff
@@ -76,21 +72,56 @@ def get_data(filters):
...
+def get_department_leave_approver_map(department=None):
+	conditions=''
+	if department:
+		conditions='and department_name = %(department)s or parent_department = %(department)s'%{'department': department}
+
+	# get current department and all its child
+	department_list = frappe.db.sql_list(''' SELECT name FROM `tabDepartment` WHERE disabled=0{0}'''.format(conditions)) #nosec
+
+	# retrieve approvers list from current department and from its subsequent child departments
+	return frappe._dict(frappe.db.sql('''
```
Comment (surajshetty3416) [inline erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py]: Use ORM.

For child departments use `get_descendants`

**Review by surajshetty3416 — CHANGES_REQUESTED** [review]

File: erpnext/hr/report/employee_leave_balance/employee_leave_balance.py [hunk 2]
```diff
@@ -54,9 +56,11 @@ def get_data(filters, leave_types):
 		filters=conditions,
 		fields=["name", "employee_name", "department", "user_id"])
 
+	department_approver_map = get_department_leave_approver_map(filters.get('department'))
+
 	data = []
 	for employee in active_employees:
-		leave_approvers = get_approvers(employee.department)
+		leave_approvers = department_approver_map.get(employee.department_name, [])
```
Comment (nabinhait) [inline erpnext/hr/report/employee_leave_balance/employee_leave_balance.py]: Also consider `leave_approver` set on Employee master (introduced last week).

File: erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py [hunk 3]
```diff
@@ -76,21 +72,56 @@ def get_data(filters):
 			'leave_type': leave_type
 		})
 		for employee in active_employees:
-			row = frappe._dict({
-				'employee': employee.name,
-				'employee_name': employee.employee_name
-			})
+			leave_approvers = department_approver_map.get(employee.department_name, [])
+
+			if (len(leave_approvers) and user in leave_approvers) or (user in ["Administrator", employee.user_id]) \
```
Comment (nabinhait) [inline erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py:78]: Also consider leave_approver set on Employee master (introduced last week).

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/hr/report/employee_leave_balance/employee_leave_balance.py [hunk 4]
```diff
@@ -52,11 +54,13 @@ def get_data(filters, leave_types):
...
 		filters=conditions,
-		fields=["name", "employee_name", "department", "user_id"])
+		fields=["name", "employee_name", "department", "user_id", "leave_approver"])
+
+	department_approver_map = get_department_leave_approver_map(filters.get('department'))
 
 	data = []
 	for employee in active_employees:
-		leave_approvers = get_approvers(employee.department)
+		leave_approvers = employee.leave_approver or department_approver_map.get(employee.department_name, [])
```
Comment (nabinhait) [inline erpnext/hr/report/employee_leave_balance/employee_leave_balance.py]: Leave Approver mentioned in Employee is default approver. All the other approvers for the department should also be able to view the report.

You should just add employee.leave_approver to department_leave_approvers list, if not exists.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py [hunk 5]
```diff
@@ -76,21 +72,59 @@ def get_data(filters):
...
+
+	return conditions
+
+def get_department_leave_approver_map(department=None):
+	conditions=''
+	if department:
+		conditions='and department_name = %(department)s or parent_department = %(department)s'%{'department': department}
+
+	# get current department and all its child
+	department_list = frappe.db.sql_list(''' SELECT name FROM `tabDepartment` WHERE disabled=0{0}'''.format(conditions)) #nosec
```
Comment (surajshetty3416) [inline erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py]: This also can be converted to ORM. Why raw query?

**Review by surajshetty3416 — CHANGES_REQUESTED** [review]

File: erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py (hunk 5 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/report/employee_leave_balance_summary/employee_leave_balance_summary.py]: no equivalent ORM with similar functionality

File: erpnext/hr/report/employee_leave_balance/employee_leave_balance.py (hunk 4 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/report/employee_leave_balance/employee_leave_balance.py]: I assumed leave approver in employee was overriding leave approver in the department. Added corrected changes.

**Review by nabinhait — APPROVED** [review]

**Comment by muchai** [comment]
An error on these two reports: https://github.com/frappe/erpnext/issues/19100

**Pull request merged**